docs/decisions/0013-library-reference-content-block.rst

```diff
@@ -0,0 +1,58 @@
+Referencing Content Blocks in Library V2
+--------------------------------------------------
+
+Status
+=======
+Pending
+
+Context
+=======
+Currently, the courses are being stored in the Modulestore and we want the library content to be stored in the Blockstore.
+To achieve this we need a referencing mechanism to reference the libraries to the courses. This referencing helps in replacing
+v1 libraries (current content libraries stored on Modulestore) with v2 libraries (Blockstore backed) which are:
+
+Currently, courses are stored in Modulestore (MongoDB), and libraries can be stored in either Modulestore (libraries v1) or blockstore (libraries v2).
+We are currently building out the functionality of libraries v2, and trying to migrate all usage of content libraries onto Blockstore.
+The benefits of using the Blockstore backed v2 libraries are:
+
+#. Blockstore is much more flexible than modulestore and can store anything that can be represented as a file - unlike modulestore, which only stores courses built of XBlocks. Also, the Blockstore is being moved into the edx-platform.
+#. simpler to maintain than modulestore
+#. increases re-usability of the content
+
+Terminology Used
+^^^^^^^^^^^^^^^^
+#. Blocks: Each reusable piece of content in a library. Example: problems, videos, HTML. Also known as Xblocks.
+#. Library: Collection of blocks in a course.
+
+
+Decisions
+=========
+We want to be able to reference single and multiple specific blocks from a library (v2, as well as v1) into a course (that are stored in modulestore)
+The workflow discussed for the interface for library content referencing would be:
+
+#. Author chooses a library.
+#. Author chooses a pool of blocks.
+#. If multiple blocks of the same type are selected, ask for randomization
+#. If no randomization, then ask for ordering
+
+
+Implementation
+^^^^^^^^^^^^^^
+The library referencing flow would be implemented within the Studio and/or LMS processes of a single Open edX instance.
+The library content would be stored and rendered outside the modulestore.
+
+We would achieve this by introducing a unit_compositor subsystem. Like learning_sequences, the subsystem would be populated by CMS upon course publish. It would store a read-optimized form of:
+* Metadata and child-block lists for all course Units, and
+* Definitions of library blocks.
+
+We would then update the LmsXBlockRuntime (called “CombinedSystem” until BD-13 is done) to use the unit_compositor as its backing store for units. When a unit is requested for a particular user, the unit_compositor would:
+#. Load the unit’s child blocks from modulestore.
```

Comment:
> To achieve this we need a referencing mechanism to reference the libraries to the courses. This referencing helps in replacing v1 libraries (current content libraries stored on Modulestore) with v2 libraries (Blockstore backed) which are:
>
> Currently, courses are stored in Modulestore (MongoDB), and libraries can be stored in either Modulestore (libraries v1) or blockstore (libraries v2). We are currently building out the functionality of libraries v2, and trying to migrate all usage of content libraries onto Blockstore.

I'm a bit confused here. Above it is mentioned that referencing libraries could happen from two stores: `blockstore` and `modulestore`, but this implementation only mentions referencing libraries from `modulestore`.

Also, based on @bradenmacdonald's [comment](https://github.com/edx/edx-platform/pull/29633/files#r771759493) it seems that this implementation is the opposite of what `LibrarySourcedBlock` does. Shouldn't this implementation has functionality to reference libraries from both `block-` and `modulestore`?

Reply:
I agree it should be more clear, but this makes sense to me.

Courses are always stored in modulestore, and the first thing this does is load the **course unit** from modulestore. Then, if that course unit happens to contain these new "library reference blocks", it will replace each "library reference block" (which was part of the course, so loaded from modulestore) with the specified library XBlocks, which would be loaded from Blockstore.

> it seems that this implementation is the opposite of what LibrarySourcedBlock does. Shouldn't this implementation has functionality to reference libraries from both block- and modulestore?

No, this is doing more or less the same thing (use blockstore library blocks in a course), it just does it in a different way (uses a just-in-time compositor instead of copying the blocks in advance)